ClientGUI/IME/IMEHandler.cs

```diff
@@ -0,0 +1,215 @@
+
+    private void OnCompositionChanged(string oldValue, string newValue)
+    {
+        Debug.WriteLine($"IME: OnCompositionChanged: {newValue.Length - oldValue.Length}");
+
+        IMEEventReceived = true;
+        // It seems that OnIMETextInput() is always triggered after OnCompositionChanged(). We expect such a behavior.
+        LastActionIMEChatInput = false;
+    }
+
+    protected ConcurrentDictionary<XNATextBox, Action<char>?> TextBoxHandleChatInputCallbacks = [];
+
+    public virtual int CompositionCursorPosition { get; set; }
+
+    public static IMEHandler Create(Game game)
+    {
+#if DX
+        return new WinFormsIMEHandler(game);
+#elif XNA
+        // Warning: Think carefully before enabling WinFormsIMEHandler for XNA builds!
+        // It *might* occasionally crash due to an unknown stack overflow issue.
+        // This *might* be caused by both ImeSharp and XNAUI hooking into WndProc.
+        // ImeSharp: https://github.com/ryancheung/ImeSharp/blob/dc2243beff9ef48eb37e398c506c905c965f8e68/ImeSharp/InputMethod.cs#L170
+        // XNAUI: https://github.com/Rampastring/Rampastring.XNAUI/blob/9a7d5bb3e47ea50286ee05073d0a6723bc6d764d/Input/KeyboardEventInput.cs#L79
+        //
+        // That said, you can try returning a WinFormsIMEHandler and test if it is stable enough now. Who knows?
+        return new DummyIMEHandler();
+#elif GL
+        return new SdlIMEHandler(game);
+#else
+#error Unknown variant
+#endif
+    }
+
+    public abstract void SetTextInputRectangle(Rectangle rectangle);
+
+    public abstract void StartTextComposition();
+
+    public abstract void StopTextComposition();
+
+    protected virtual void OnIMETextInput(char character)
+    {
+        Debug.WriteLine($"IME: OnIMETextInput: {character} {(short)character}; IMEFocus is null? {IMEFocus == null}");
+
+        IMEEventReceived = true;
+        LastActionIMEChatInput = true;
+
+        if (IMEFocus != null)
+        {
+            TextBoxHandleChatInputCallbacks.TryGetValue(IMEFocus, out var handleChatInput);
+            handleChatInput?.Invoke(character);
+        }
+    }
+
+    private void SetIMETextInputRectangle(XNATextBox sender)
+    {
+        WindowManager windowManager = sender.WindowManager;
+
+        Rectangle textBoxRect = sender.RenderRectangle();
+        double scaleRatio = windowManager.ScaleRatio;
+
+        Rectangle rect = new()
+        {
+            X = (int)(textBoxRect.X * scaleRatio + windowManager.SceneXPosition),
+            Y = (int)(textBoxRect.Y * scaleRatio + windowManager.SceneYPosition),
+            Width = (int)(textBoxRect.Width * scaleRatio),
+            Height = (int)(textBoxRect.Height * scaleRatio)
+        };
+
+        // The following code returns a more accurate location, aware of the input cursor
+        // However, it requires SetIMETextInputRectangle() be called whenever InputPosition is changed.
+        // And therefore, it's commented out for now.
```

Comment:
Not clear why you can't just call that

Reply:
Comment updated.
If we apply this code, sometimes it's precise, other times it's worse than the inaccurate codes (due to outdated InputPosition). Such an inconsistent behavior is worse than a consistent but manageable inaccuracy.